File: Data Structures & Algorithms/max-area-of-island/submission-1.py
DIRECTIONS = [(-1, 0), (1, 0), (0, 1), (0, -1)]
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])

        visit = set()

        def dfs(r, c):
            if r < 0 or r >= m or c < 0 or c >= n or grid[r][c] == 0 or (r, c) in visit:
                return 0
            visit.add((r,c))
            ans = 1
            for dr, dc in DIRECTIONS:
                ans += dfs(r + dr, c+dc)
            return ans

        area = 0
        for r in range(m):
            for c in range(n):
                area = max(area, dfs(r, c))
        return area                
                

File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {
    public int numIslands(char[][] grid) {
        int row = grid.length;
        int col = grid[0].length;
        int ans = 0;

        for (int r = 0; r < row; r++) {
            for (int c = 0; c < col; c++) {
                if (grid[r][c] == '1') {
                    ans++;
                    dfs(grid, r, c);
                }
            }
        }
        return ans;
    }

    private void dfs(char[][] grid, int r, int c) {
        int row = grid.length;
        int col = grid[0].length;
        if (r < 0 || r >= row || c < 0 || c >= col || grid[r][c] == '0') {
            return;
        }

        grid[r][c] = '0';

        dfs(grid, r - 1, c);
        dfs(grid, r + 1, c);
        dfs(grid, r, c - 1);
        dfs(grid, r, c + 1);
    }
}
